### PR TITLE
[redux-pack]: add key signature to ActionMeta interface

### PR DESCRIPTION
**Pull request by no-stack-dub-sack**
@tansongyang @andy-ms This is a PR to add a key signature to the `ActionMeta` interface of the redux-pack module. This will allow redux-pack / TypeScript users to pass custom metadata through their actions beyond just the allowed keys. There are valid use-cases for this, however the current typings do not allow it. [This line](https://github.com/lelandrichardson/redux-pack/blob/7818ffd4304d5f0e2c94056f3626a399fc9a5a10/src/middleware.js#L44) of the redux-pack source code shows any keys passed through the `meta` object will be available in the returned action after passing through the middleware, not just keys that share the shape of the state object or are meta handlers.
 
As this is a very minor change, I did not see a need to modify the current tests.

**NOTE**: `npm run lint redux-pack` was failing on my machine before adding the key signature but everything seems to pass CI fine:
image

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://github.com/lelandrichardson/redux-pack/blob/7818ffd4304d5f0e2c94056f3626a399fc9a5a10/src/middleware.js#L44>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by tansongyang** [comment]
@no-stack-dub-sack Thanks for the PR. I don't really do much work in React-land these days, so it would help me out if you could post a snippet of code that you are now able to write but couldn't write before these changes.

In fact, I would prefer if you update or add a new test. I know it's a small change, but having tests in the diff will help when reading through the history.

**Review by tansongyang — CHANGES_REQUESTED** [review]
@no-stack-dub-sack Thanks for the PR. I don't really do much work in React-land these days, so it would help me if you add a test. It should demonstrate what code you are now able to write but couldn't write before these changes. I know it's a small change, but having tests in the diff will help in the future when reading through the history.

**Comment by no-stack-dub-sack** [comment]
@tansongyang Sure, part of the reason I didn't do it last night was because I was on a machine without full TS support (hence the lint failing, I think), and I would have been unable to verify the tests :-)

I'm back on my usual machine now and should be able to add something. Thanks for the quick reply.

**Comment by no-stack-dub-sack** [comment]
@tansongyang hmmm... now that I'm taking a deeper look at this and trying to write tests, I think there may be a larger issue, which I think is multifold. I'll try to explain here...

The extended `Action` type only allows for actions provided by redux-pack, however, what if a reducer has a non-async action that does not rely on the `handle` helper, and/or responds to an action whose shape does not conform to:

```js
{
  type: 'ACTION_TYPE',
  payload,
  meta
}
```

For example:

```js
import { handle, Action } from 'redux-pack';

interface IState {
  isLoading: boolean;
  fooError: string | null;
  tooltip: string;
  foo: {
    id: number;
    phrase: string;
  }
};

export function fooReducer(state: IState = initialState, action: Action<IState>) {
  const { type, payload } = action;
  switch (type) {
    case UPDATE_FOO:
      return {
        ...state,
        foo: {
          ...state.foo,
          phrase: action.newPhrase
          // [ts] Property 'newPhrase' does not exist on type 'Action<IState, {}, {}, {}>'.
        }
      };
    case LOAD_FOO:
      return handle(state, action, {
        start: prevState => ({
          ...prevState,
          isLoading: true,
          fooError: null
        }),
        finish: prevState => ({ ...prevState, isLoading: false }),
        failure: prevState => ({ ...prevState, fooError: payload }),
        success: prevState => ({ ...prevState, foo: payload }),
      });
    default:
      return state;
  }
}
```

For reducers responding to multiple actions, a more common pattern is to define a type `Actions` which is [a Union Type of all available actions](https://github.com/fbsamples/f8app/blob/master/js/actions/types.js), however this pattern is not really usable if you are typing your action parameter this way (adding `Action` from redux-pack typings to the union creates new problems of its own).

Furthermore, the original use case I submitted this PR for becomes impossible, i.e., passing along custom metadata through the middleware and having access to it in the reducer case handling that action (which is allowable by the library):

```js
const loadFoo= (tooltip: string, query: string): LoadFooAction=> {
  return {
    type: 'LOAD_FOO',
    promise: axios.get(`/api/foo?query=${query}`),
    meta: {
      tooltip
    }
  }
}
```

```js
    case LOAD_FOO:
      return handle(state, action, {
      // [ts] Types of property 'meta' are incompatible.
      // Type '{ tooltip: string; }' is not assignable to type 'ActionMeta<{ isLoading: boolean; fooError: string | null; tooltip: string; foo: {...'.
      // Type '{ toolstip: string; }' has no properties in common with type 'ActionMeta<{ isLoading: boolean; fooError: string | null; tooltip: string; foo: {...'.
        start: prevState => ({
          ...prevState,
          isLoading: true,
          fooError: null
        }),
        finish: prevState => ({ ...prevState, isLoading: false }),
        failure: prevState => ({ ...prevState, fooError: payload }),
        success: prevState => ({ ...prevState, foo: payload, tooltip: action.meta.tooltip }),
      });
```

Does this make sense at all? It's a bit hard to explain, and my apologies if I'm missing something here, but I think the issue it with explicitly typing the `action` parameter with `Action` from redux-pack typings.

For the above use case, in my actual code, I had `action` typed with a Union type of all reducer actions as described above, and adding a key signature to the `ActionMeta` fixed the issue, and redux-pack allowed it. Here is how the action (as part of the Union) was typed:

```js
export interface FetchEntitiesAction {
  type: 'FETCH_ENTITIES';
  promise: AxiosPromise<IResult<Entity>>;
  payload?: AxiosResponse<IResult<Entity>>;
  meta: {
    system: Systems;
  }
}
```

Even though this was present in the union, `action` inside `handle` threw the error, as `meta.system` was not a recognized property. Hence the solution was to add the key signature. However, now, when attempting to add tests, I see that there is something a bit bigger going on here. With the current typings, and actually using `Action` to type  the reducer action, I then got the error that `meta` might be `undefined` since this is an optional property. If a user is using a Union type to type their actions, adding the key-signature is a viable solution, however if they are using `Action` from redux-pack, it is still problematic, and additionally, they will not be allowed to have an action that contains keys other than `type`, `payload` and `meta` in the same reducer.

Hope this makes sense... but as someone fairly new to typescript, and completely new to writing declaration files, I'm not sure what the proper solution will be that will work across all use-cases.

**Comment by no-stack-dub-sack** [comment]
Here are the tests I started writing which describes part of the issue:

```js
interface Bar {
  id: string;
  dateAdded: Date;
  barInfo: string;
}
declare const BarApi: {
  get(query: string): Promise<Bar[]>;
};

const FETCH_BAR = 'FETCH_BAR';
function fetchBar(query: string, tooltip: string): Action<BarState, Bar[]> {
  return {
    type: FETCH_BAR,
    promise: BarApi.get(query),
    meta: {
      tooltip
      // a user should be able to pass custom metadata through to their reducer
    }
  }
}

interface BarState {
  error: string | null;
  isLoading: boolean;
  tooltip: string;
  recordset: (Bar | void)[];
}

declare const barInitialState: BarState;
function barReducer(state = barInitialState, action: Action<BarState, Bar, string | null>) {
  const { type, payload } = action;
  switch (type) {
    case FETCH_BAR:
      return handle(state, action, {
        start: prevState => ({ ...prevState, isLoading: true, error: null, recordset: [] }),
        finish: prevState => ({ ...prevState, isLoading: false }),
        failure: prevState => ({ ...prevState, error: `${type} error!` }),
        success: prevState => ({ ...prevState, recordset: payload as Bar[], tooltip: action.meta.tooltip }), // [ts] Object is possibly 'undefined'.
      });
    default:
      return state;
  }
}
```
Getting close, but fails with strict null checks because `meta` is an optional property (possibly `strictNullCheck` could be disabled to allow for this behavior?

**Comment by no-stack-dub-sack** [comment]
@tansongyang Sure, no worries at all, please let me know if you need clarification on anything.

**Comment by no-stack-dub-sack** [comment]
@tansongyang so, I've been trying to do a bit more research on this, because I figure this may be a common issue (middleware returning actions that don't match originally defined actions and how to get around that), and I turned up an article that actually specifically mentions redux-pack that may help shed some light on this.

There's a part that talks about the tagged unions pattern I mentioned, but more specifically, the [Typed Middleware](https://medium.com/p/1b2dc2ed4790#f587) section addresses redux-pack.

**Comment by tansongyang** [comment]
@no-stack-dub-sack Would you mind updating this PR with your proposed changes so I can see how it looks? You might also consider adding yourself as a maintainer, and maybe even removing me from as a maintainer. As I said, I don't work much in this space anymore, and so I'm a lot slower with handling PRs. However, I think if you took over, this package would be in very capable hands.

**Comment by no-stack-dub-sack** [comment]
@tansongyang well, the changes I'm now suggesting, aren't actually my changes, the new type-defs were written by the creator of redux-guards, @quicksnap. That said, he did mention in his comment that he's been too busy to actually make an upstream PR to merge them in, so I'm happy to modify this PR to include his changes.

I realized that the reason it's now working for me, is not only because of the redux-guards pattern, but also because that package is pulling in his updated  typings for redux-pack. Looking at the declaration file more closely, I think it would address the root cause of this issue as well, but it would require a bit more testing.

Lastly, regarding the maintainer bit:  while I'd love the OSS cred, I'm still fairly new to TypeScript, and I can't honestly say I understand the advanced concepts found in these typings well enough to be able to answer questions from other contributors, so I may not be the best person to take over at this point. 

Maybe I'll feel more confident after working through this PR and can circle back on that. Let me try and make some changes tonight and see what I can come up with.

**Comment by quicksnap** [comment]
For reference, here are the forked typings that I use in our project: https://gist.github.com/quicksnap/06e0d46d5c30510d5f65276974d3911b

As mentioned, I haven't had the time to invest in making sure these typings are totally accurate, but they better describe how redux-pack works in my situation.

File: types/redux-pack/index.d.ts
Comment (no-stack-dub-sack) [inline types/redux-pack/index.d.ts]: @tansongyang the addition of the `TMetaPayload` argument to the Action generic and the intersection typing of `ActionMeta` and `TMetaPayload` on this line (along with @quicksnap's new `PackActionPayload` and `MetaPayload` typings) seem to correctly solve the original use case of passing custom metadata keys through to the reducer (without adding the key signature to the `ActionMeta` type def as I originally did. This seems like the safer and better solution.

File: types/redux-pack/redux-pack-tests.ts
Comment (no-stack-dub-sack) [inline types/redux-pack/redux-pack-tests.ts]: :point_up: accessing of custom meta props did not work with previous typings

File: types/redux-pack/index.d.ts
Comment (no-stack-dub-sack) [inline types/redux-pack/index.d.ts]: `TFullState` was originally a param of the handle generic, but as it was only used once in the def, the TSLint rules forced me to define it like this, which I dont thin should be a problem as it allows for an obj of any shape.

File: types/redux-pack/index.d.ts
Comment (no-stack-dub-sack) [inline types/redux-pack/index.d.ts]: see note on line 68

**Comment by no-stack-dub-sack** [comment]
@tansongyang ok, incorporated @quicksnap's improved typings with some slight  modifications to solve the original use case without the redux-guards pattern. Slight other changes, all describe above. Modified the tests to reflect these changes. 

@quicksnap, if you have the time, if you'd like to review this as well, that could be beneficial, just to make sure the additions / modifications I made to your work seem to fit.  

Let me know what you think... can squash these commits when ready to merge.

File: types/redux-pack/index.d.ts
Comment (tansongyang) [inline types/redux-pack/index.d.ts]: We should increment the version here to `0.2`.

**Review by tansongyang — CHANGES_REQUESTED** [review]
Thanks for the hard work @no-stack-dub-sack  and @quicksnap! I'm fine with these changes going out. I just want to increment the version number, since we've significantly changed the types.

**Comment by no-stack-dub-sack** [comment]
@tansongyang I thought of that as well, and when I researched it, it looks like that should always refer to the version of the lib itself: https://definitelytyped.org/guides/contributing.html

This is a little outdated and not totally clear though.

Also, I do feel much more comfortable with this declaration file now, so your call about the maintainer thing. I'm happy to at least be added and take the lead, but might be nice to have a backup if something pops up I can't handle (not that I expect a ton of activity anyway)

**Review by tansongyang — APPROVED** [review]
@no-stack-dub-sack Ah yes, you're right about the versioning. Thanks again for the contribution! Glad to have you.

**Comment by no-stack-dub-sack** [comment]
@tansongyang Happy to help, thanks for the review!

**Pull request merged**
